### PR TITLE
http: improving shadowing with upstream filters

### PR DESCRIPTION
**Pull request by alyssawilk**
Fixing a bug where the router was not properly shut down for shadowing with upstream filters.
Fixing a second bug where headers were latched too late, and request headers were not isolated from shadow headers

Testing: new integration tests
Docs Changes: n/a
Release Notes: n/a
Part of https://github.com/envoyproxy/envoy/issues/10455

### CODE REVIEW

**Comment by alyssawilk** [comment]
cc @paul-r-gall this is a precursor to fixing up headers for upstream filters + shadowing.

**Comment by alyssawilk** [comment]
cc @paul-r-gall ok, now has the header fixes in.
Also you note I had to tweak the add body filter to no longer reference the downstream connection - it was nullptr for the mirrored requests.  you possibly leverage that to not run encode functions for your mirrored requests?

**Review by mattklein123 — APPROVED** [review]
Thanks, LGTM!

File: source/common/http/async_client_impl.h [hunk 1]
```diff
@@ -93,6 +93,7 @@ class AsyncStreamImpl : public AsyncClient::Stream,
 public:
   AsyncStreamImpl(AsyncClientImpl& parent, AsyncClient::StreamCallbacks& callbacks,
                   const AsyncClient::StreamOptions& options);
+  ~AsyncStreamImpl() override { router_.onDestroy(); }
```
Comment (mergeconflict) [inline source/common/http/async_client_impl.h:96]: Sorry for the late drive-by here, but quick question: would it make sense for `Router::ProdFilter` or `Router::Filter`'s destructor to call its own `onDestroy`? (I'm looking at the comments on `StreamFilterBase::onDestroy` but I don't really understand whether they're applicable here.)
